samples/subsys/usb/hid-mouse/src/main.c

```diff
@@ -1,245 +1,100 @@
 
 static void status_cb(enum usb_dc_status_code status, const uint8_t *param)
 {
 	usb_status = status;
 }
 
-static void left_button(const struct device *gpio, struct gpio_callback *cb,
-			uint32_t pins)
+static void ALWAYS_INLINE rwup_if_suspended(void)
 {
-	int ret;
-	uint8_t state = status[MOUSE_BTN_REPORT_POS];
-
 	if (IS_ENABLED(CONFIG_USB_DEVICE_REMOTE_WAKEUP)) {
 		if (usb_status == USB_DC_SUSPEND) {
 			usb_wakeup_request();
 			return;
 		}
 	}
-
-	ret = gpio_pin_get(gpio, sw0.pin);
-	if (ret < 0) {
-		LOG_ERR("Failed to get the state of port %s pin %u, error: %d",
-			gpio->name, sw0.pin, ret);
-		return;
-	}
-
-	if (def_val[0] != (uint8_t)ret) {
-		state |= MOUSE_BTN_LEFT;
-	} else {
-		state &= ~MOUSE_BTN_LEFT;
-	}
-
-	if (status[MOUSE_BTN_REPORT_POS] != state) {
-		status[MOUSE_BTN_REPORT_POS] = state;
-		k_sem_give(&sem);
-	}
 }
 
-static void right_button(const struct device *gpio, struct gpio_callback *cb,
-			 uint32_t pins)
+static void input_cb(struct input_event *evt)
 {
-	int ret;
-	uint8_t state = status[MOUSE_BTN_REPORT_POS];
-
-	if (IS_ENABLED(CONFIG_USB_DEVICE_REMOTE_WAKEUP)) {
-		if (usb_status == USB_DC_SUSPEND) {
-			usb_wakeup_request();
-			return;
+	uint8_t tmp[MOUSE_REPORT_COUNT];
+
+	(void)memcpy(tmp, report, sizeof(tmp));
+
+	switch (evt->code) {
+	case INPUT_KEY_0:
+		rwup_if_suspended();
+		WRITE_BIT(tmp[MOUSE_BTN_REPORT_IDX], MOUSE_BTN_LEFT, evt->value);
+		break;
+	case INPUT_KEY_1:
+		rwup_if_suspended();
+		WRITE_BIT(tmp[MOUSE_BTN_REPORT_IDX], MOUSE_BTN_RIGHT, evt->value);
+		break;
+	case INPUT_KEY_2:
+		if (evt->value) {
+			tmp[MOUSE_X_REPORT_IDX] += 10U;
 		}
-	}
-
-	ret = gpio_pin_get(gpio, sw1.pin);
-	if (ret < 0) {
-		LOG_ERR("Failed to get the state of port %s pin %u, error: %d",
-			gpio->name, sw1.pin, ret);
-		return;
-	}
-
-	if (def_val[1] != (uint8_t)ret) {
-		state |= MOUSE_BTN_RIGHT;
-	} else {
-		state &= ~MOUSE_BTN_RIGHT;
-	}
-
-	if (status[MOUSE_BTN_REPORT_POS] != state) {
-		status[MOUSE_BTN_REPORT_POS] = state;
-		k_sem_give(&sem);
-	}
-}
-
-static void x_move(const struct device *gpio, struct gpio_callback *cb,
-		   uint32_t pins)
-{
-	int ret;
-	uint8_t state = status[MOUSE_X_REPORT_POS];
-
-	ret = gpio_pin_get(gpio, sw2.pin);
-	if (ret < 0) {
-		LOG_ERR("Failed to get the state of port %s pin %u, error: %d",
-			gpio->name, sw2.pin, ret);
-		return;
-	}
-
-	if (def_val[2] != (uint8_t)ret) {
-		state += 10U;
-	}
 
-	if (status[MOUSE_X_REPORT_POS] != state) {
-		status[MOUSE_X_REPORT_POS] = state;
-		k_sem_give(&sem);
-	}
-}
-
-static void y_move(const struct device *gpio, struct gpio_callback *cb,
-		   uint32_t pins)
-{
-	int ret;
-	uint8_t state = status[MOUSE_Y_REPORT_POS];
+		break;
+	case INPUT_KEY_3:
+		if (evt->value) {
+			tmp[MOUSE_Y_REPORT_IDX] += 10U;
+		}
 
-	ret = gpio_pin_get(gpio, sw3.pin);
-	if (ret < 0) {
-		LOG_ERR("Failed to get the state of port %s pin %u, error: %d",
-			gpio->name, sw3.pin, ret);
+		break;
+	default:
 		return;
```

Comment:
Should there be a LOG_INF for "unrecognized code" here? May help out if someone is trying this on a custom board with different codes.